From 7a472fcab2c5ab6c5878600094af7d1ae8ddb4ba Mon Sep 17 00:00:00 2001
From: Vlad Zahorodnii <vlad.zahorodnii@kde.org>
Date: Wed, 17 Jul 2024 14:57:40 +0300
Subject: [PATCH] backends/x11: Fix crash that happens when toggling
 compositing

When the GLX or the EGL backend is destroyed, it is going to reset the
RenderLoop state, including the number of frames in flight. It does so
because of the historical reasons. At the time, there was no output frame
object to track the lifecycle of a frame.

After introducing the OutputFrame and hooking it into the RenderLoop,
the pending frame count will be reset automatically in RenderLoop when
the GLX or the EGL backend is destroyed. But we forgot to remove
the invalidate() function calls. So, when the GLX backend goes down, it
resets the pending frame count to zero and then it destroys the pending
OutputFrame object, which would result in decreasing the pending frame
count by 1 and triggering an assert in the RenderLoopPrivate::notifyFrameDropped()
function.

Since there is the OutputFrame helper now, the invalidate() function
can be dropped. Technically, the invalidate function did more than just
reset pendingFrameCount to 0, for example also stop the compositeTime.
But that should be fairly harmless new behavior.
---
 src/backends/x11/standalone/x11_standalone_egl_backend.cpp | 6 ------
 src/backends/x11/standalone/x11_standalone_glx_backend.cpp | 5 -----
 src/core/renderloop.cpp                                    | 7 -------
 src/core/renderloop_p.h                                    | 1 -
 4 files changed, 19 deletions(-)

diff --git a/src/backends/x11/standalone/x11_standalone_egl_backend.cpp b/src/backends/x11/standalone/x11_standalone_egl_backend.cpp
index 5c980532383..60fe6b7ab1a 100644
--- a/src/backends/x11/standalone/x11_standalone_egl_backend.cpp
+++ b/src/backends/x11/standalone/x11_standalone_egl_backend.cpp
@@ -10,7 +10,6 @@
 #include "core/outputbackend.h"
 #include "core/outputlayer.h"
 #include "core/overlaywindow.h"
-#include "core/renderloop_p.h"
 #include "opengl/eglcontext.h"
 #include "opengl/egldisplay.h"
 #include "opengl/glplatform.h"
@@ -79,11 +78,6 @@ EglBackend::EglBackend(::Display *display, X11StandaloneBackend *backend)
 
 EglBackend::~EglBackend()
 {
-    // No completion events will be received for in-flight frames, this may lock the
-    // render loop. We need to ensure that the render loop is back to its initial state
-    // if the render backend is about to be destroyed.
-    RenderLoopPrivate::get(m_backend->renderLoop())->invalidate();
-
     m_query.reset();
 
     if (isFailed() && m_overlayWindow) {
diff --git a/src/backends/x11/standalone/x11_standalone_glx_backend.cpp b/src/backends/x11/standalone/x11_standalone_glx_backend.cpp
index 0f646778bdb..46a0d72f363 100644
--- a/src/backends/x11/standalone/x11_standalone_glx_backend.cpp
+++ b/src/backends/x11/standalone/x11_standalone_glx_backend.cpp
@@ -27,7 +27,6 @@
 #include "compositor.h"
 #include "core/outputbackend.h"
 #include "core/overlaywindow.h"
-#include "core/renderloop_p.h"
 #include "opengl/glrendertimequery.h"
 #include "options.h"
 #include "scene/surfaceitem_x11.h"
@@ -149,10 +148,6 @@ GlxBackend::GlxBackend(::Display *display, X11StandaloneBackend *backend)
 GlxBackend::~GlxBackend()
 {
     m_vsyncMonitor.reset();
-    // No completion events will be received for in-flight frames, this may lock the
-    // render loop. We need to ensure that the render loop is back to its initial state
-    // if the render backend is about to be destroyed.
-    RenderLoopPrivate::get(m_backend->renderLoop())->invalidate();
 
     m_query.reset();
 
diff --git a/src/core/renderloop.cpp b/src/core/renderloop.cpp
index 0c6cd59175f..b42376a2620 100644
--- a/src/core/renderloop.cpp
+++ b/src/core/renderloop.cpp
@@ -177,13 +177,6 @@ void RenderLoopPrivate::dispatch()
     pendingRepaint = false;
 }
 
-void RenderLoopPrivate::invalidate()
-{
-    pendingReschedule = false;
-    pendingFrameCount = 0;
-    compositeTimer.stop();
-}
-
 RenderLoop::RenderLoop(Output *output)
     : d(std::make_unique<RenderLoopPrivate>(this, output))
 {
diff --git a/src/core/renderloop_p.h b/src/core/renderloop_p.h
index ce4d528ab01..757946e46c2 100644
--- a/src/core/renderloop_p.h
+++ b/src/core/renderloop_p.h
@@ -28,7 +28,6 @@ public:
     explicit RenderLoopPrivate(RenderLoop *q, Output *output);
 
     void dispatch();
-    void invalidate();
 
     void delayScheduleRepaint();
     void scheduleNextRepaint();
-- 
GitLab

